src/components/views/rooms/MessageComposerInput.js

```diff
@@ -1018,167 +1287,237 @@ export default class MessageComposerInput extends React.Component {
+        // providers only search for the first match which intersects with the current selection.
+        // This avoids us having to serialize the whole thing to plaintext and convert
+        // selection offsets in & out of the plaintext domain.
 
-    getAutocompleteQuery(contentState: ContentState) {
-        // Don't send markdown links to the autocompleter
-        return this.removeMDLinks(contentState, ['@', '#']);
+        if (editorState.selection.anchorKey) {
+            return editorState.document.getDescendant(editorState.selection.anchorKey).text;
+        }
+        else {
+            return '';
+        }
     }
 
-    removeMDLinks(contentState: ContentState, prefixes: string[]) {
-        const plaintext = contentState.getPlainText();
-        if (!plaintext) return '';
-        return plaintext.replace(REGEX_MATRIXTO_MARKDOWN_GLOBAL,
-        (markdownLink, text, resource, prefix, offset) => {
-            if (!prefixes.includes(prefix)) return markdownLink;
-            // Calculate the offset relative to the current block that the offset is in
-            let sum = 0;
-            const blocks = contentState.getBlocksAsArray();
-            let block;
-            for (let i = 0; i < blocks.length; i++) {
-                block = blocks[i];
-                sum += block.getLength();
-                if (sum > offset) {
-                    sum -= block.getLength();
-                    break;
-                }
-            }
-            offset -= sum;
-
-            const entityKey = block.getEntityAt(offset);
-            const entity = entityKey ? contentState.getEntity(entityKey) : null;
-            if (entity && entity.getData().isCompletion) {
-                // This is a completed mention, so do not insert MD link, just text
-                return text;
-            } else {
-                // This is either a MD link that was typed into the composer or another
-                // type of pill (e.g. room pill)
-                return markdownLink;
-            }
-        });
+    getSelectionRange(editorState: Value) {
+        let beginning = false;
+        const query = this.getAutocompleteQuery(editorState);
+        const firstChild = editorState.document.nodes.get(0);
+        const firstGrandChild = firstChild && firstChild.nodes.get(0);
+        beginning = (firstChild && firstGrandChild &&
+                     firstChild.object === 'block' && firstGrandChild.object === 'text' &&
+                     editorState.selection.anchorKey === firstGrandChild.key);
+
+        // return a character range suitable for handing to an autocomplete provider.
+        // the range is relative to the anchor of the current editor selection.
+        // if the selection spans multiple blocks, then we collapse it for the calculation.
+        const range = {
+            beginning, // whether the selection is in the first block of the editor or not
+            start: editorState.selection.anchorOffset,
```

Comment:
Does this handle a selection that has an anchorOffset that is ahead of the focusOffset ?

Reply:
Oh. It does.